src/tamp_improv/benchmarks/base.py

```diff
@@ -0,0 +1,117 @@
+"""Base environment interface."""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Generic, Protocol, TypeVar
+
+import gymnasium as gym
+from relational_structs import LiftedOperator, PDDLDomain, Predicate, Type
+from task_then_motion_planning.structs import Perceiver, Skill
+
+ObsType = TypeVar("ObsType")
+ActType = TypeVar("ActType")
+
+
+class PredicateContainer(Protocol):
+    """Protocol for predicate containers."""
+
+    def __getitem__(self, key: str) -> Predicate:
+        """Get predicate by name."""
+
+    def as_set(self) -> set[Predicate]:
+        """Convert to set of predicates."""
+
+
+@dataclass
+class PlanningComponents(Generic[ObsType]):
+    """Container for all planning-related components."""
+
+    types: set[Type]
+    predicate_container: PredicateContainer
+    operators: set[LiftedOperator]
+    skills: set[Skill]
+    perceiver: Perceiver[ObsType]
+
+
+class BaseTAMPSystem(Generic[ObsType, ActType], ABC):
+    """Base class for Task-and-Motion Planning (TAMP) systems.
+
+    This class combines:
+    1. The actual environment (gym.Env) that represents the physical world
+    2. The agent's planning components (types, predicates, operators, etc.)
+       that represent the agent's abstract model of the world for planning
+    """
+
+    def __init__(
+        self,
+        planning_components: PlanningComponents[ObsType],
+        seed: int | None = None,
+    ) -> None:
+        """Initialize TAMP system.
+
+        Args:
+            planning_components: The agent's planning model/components
+            seed: Random seed for environment
+        """
+        # Create environments
+        self.env = self._create_env()
+        # Store planning components
+        self.components = planning_components
+        # Create wrapped env with components
+        self.wrapped_env = self._create_wrapped_env(planning_components)
```

Comment:
this class is looking better! but now I wonder whether a wrapped environment is something that should be included in the base system. for example, a pure TAMP approach will have no need for a wrapped environment, but it should have access to this base TAMP system right?